### ISSUE TITLE
`shuffle_after_epoch` in fn.readers.coco is using a same seed for shuffling, is it normal?

### ISSUE THREAD

**Issue by songyuc** [NVIDIA/DALI#4319]
Hi, guys,
I am learning `dali` today, and I find that, when `shuffle_after_epoch` of fn.readers.coco is used for shuffling data,
it seems to use a same seed in every program's run.
Specifically, every time I run the program,
it shows these in a first epoch,
![image](https://user-images.githubusercontent.com/27288110/194551187-8253d3c0-171e-4790-9aa3-100744845a79.png)
and it shows these two in a second epoch as
![image](https://user-images.githubusercontent.com/27288110/194551728-494b66cb-98e8-49bb-bd40-97d19299c599.png)
I don't know whether, `shuffle_after_epoch` means shuffling the data in a fixed pattern, specifically, although the images sequence is different every epoch, but all the sequences pattern are the same every run like a fixed seed.

Your answer and guide will be appreciated!

**Comment by klecki** [comment]
Correction (thanks @JanuszL for pointing that out),
our readers, when used with `shuffle_after_epoch`, use the same fixed seed: https://github.com/NVIDIA/DALI/blob/main/dali/operators/reader/loader/file_label_loader.h#L197

This is to ensure, that if you have multiple pipelines running on several devices, you still get the same order of samples between them and there is no duplication or missing samples.

The seed is used, when you specify `random_shuffle=True`. In that case, the global epoch still has the same order, but the samples are first loaded into a prefetch buffer and are randomly sampled from it - this intermediate step gives you randomness, while globally the epochs are consistent between multiple pipelines.

In theory we could allow the seed to impact the global epoch, but currently there is no mechanism in DALI which would ensure that it is consistent between multiple pipelines - and if we leave it as a default `-1` we can be almost certain that those time-based seeds won't be consistent.

**Comment by songyuc** [comment]
> Correction (thanks @JanuszL for pointing that out), our readers, when used with `shuffle_after_epoch`, use the same fixed seed: https://github.com/NVIDIA/DALI/blob/main/dali/operators/reader/loader/file_label_loader.h#L197
> 
> This is to ensure, that if you have multiple pipelines running on several devices, you still get the same order of samples between them and there is no duplication or missing samples.
> 
> The seed is used, when you specify `random_shuffle=True`. In that case, the global epoch still has the same order (you can't use it `shuffle_after_epoch`), but the samples are first loaded into a prefetch buffer and are randomly sampled from it - this intermediate step gives you randomness, while globally the epochs are consistent between multiple pipelines.
> 
> In theory we could allow the seed to impact the global epoch, but currently there is no mechanism in DALI which would ensure that it is consistent between multiple pipelines - and if we leave it as a default `-1` we can be almost certain that those time-based seeds won't be consistent.

Hi, @klecki, here is the [[script]](https://gist.github.com/songyuc/44afd768190119accdbe1e8438afc0cf) that I uses for demonstrating my observation.
It may be a little brute-force way to show, as I don't know a good way to trigger a `fn.readers.coco` to shuffle the data with `shuffle_after_epoch`.

**Comment by klecki** [comment]
Hi @songyuc,
as for now it's not possible to shuffle the whole dataset randomly in a different way for every run.
We see a way to improve the current situation by allowing you to use both `shuffle_after_epoch`  and `random_shuffle` at the same time. The first will shuffle the dataset after each epoch (but in a consistent manner between many pipelines) and the latter would randomize between samples that are picked into the batch.
We will try to look into that for DALI 1.20.

**Comment by songyuc** [comment]
> Hi @songyuc, as for now it's not possible to shuffle the whole dataset randomly in a different way for every run. We see a way to improve the current situation by allowing you to use both `shuffle_after_epoch` and `random_shuffle` at the same time. The first will shuffle the dataset after each epoch (but in a consistent manner between many pipelines) and the latter would randomize between samples that are picked into the batch. We will try to look into that for DALI 1.20.

OK, I get it.
For my better adoption of dali, I need to be very clear about the behavior of `shuffle_after_epoch`, and currently,  I am using
```python
images, bboxes, labels = fn.readers.coco(
    file_root=coco_root + "/train2017",
    annotations_file=coco_root + "/annotations_trainval2017/annotations/instances_train2017.json",
    skip_empty=True,
    ratio=True,
    ltrb=True,
    random_shuffle=False,
    shuffle_after_epoch=True,  # data shuffling
    name="Reader")
```
And I need a quick script to validate that dali do shuffle the coco data, but I find my [[script]](https://gist.github.com/songyuc/44afd768190119accdbe1e8438afc0cf) takes a long time, as it traverses the entire data set. This is partially due to that I don't know a good way to trigger `fn.readers.coco` with shuffle_after_epoch to perform shuffling action.

So could you please give some guide which help me constructing a quick script (<=10s) to make sure that my coco data will be shuffled. And I can also improve my blog post of dali 😄.

**Comment by JanuszL** [comment]
Hi @songyuc,

You can trim your pipeline to the reader only and set `image_ids=True` to return a unique sample id (according to the COCO format). Then you should be able to traverse your data set faster.

**Comment by songyuc** [comment]
Hi @JanuszL, I have modified as you suggested as [[script]](https://gist.github.com/songyuc/3759a1de255cabf4510cc692fcef80ad).
But there was an uneasy WARNING when I used `train_loader.reset()`:
> ...
29317it [00:02, 10191.16it/s]
WARNING:root:DALI iterator does not support resetting while epoch is not finished.                              Ignoring...
0it [00:00, ?it/s]len(train_loader) =  29316
...

which I guess is due to not synchronize CUDA operations.
So how should I deal with this warning?

**Comment by JanuszL** [comment]
@songyuc - the warning you see tells that the underlying DALI pipeline still has data from this epoch and issuing the `reset` won't affect the pipeline state.
In your script you don't need to call `train_loader.reset()`, the `enumerate` will call it under the hood (it will call `iter()` and this calls reset).

**Comment by songyuc** [comment]
By the way, I am turning to using `auto_reset` in `DALIGenericIterator` [[script]](https://gist.github.com/songyuc/9fc8da3ad8b851353940e054b36bc7e1), but the WARNING still existed.

**Comment by songyuc** [comment]
For better understanding, I removed `enumerate` in my loop as [[script_no_enumerate]](https://gist.github.com/songyuc/211247ffa29bdcb0525467edbf2d568f) and there was still a WARNING when using `reset()`.
According to the user-case of [[dali-torch-ssd]](https://github.com/NVIDIA/DALI/blob/main/docs/examples/use_cases/pytorch/single_stage_detector/main.py#L192-L193), I feel it essential to call `reset()` in my epochs.
So, please help me improve my demo script.

**Comment by JanuszL** [comment]
Apparently either `tqdm` calls `enumerate` under the hood or just directly calls `iter()` over the provided input. This way or another `reset()` there is not needed as well.

I think we forgot to update the mentioned SSD example. https://github.com/NVIDIA/DALI/pull/4373 should fix that.

**Comment by songyuc** [comment]
> Apparently either `tqdm` calls `enumerate` under the hood or just directly calls `iter()` over the provided input. This way or another `reset()` there is not needed as well.
> 
> I think we forgot to update the mentioned SSD example. #4373 should fix that.

Thanks sincerely for your guide!
